twill-yarn/src/main/java/org/apache/twill/internal/appmaster/RunningContainers.java

```diff
@@ -575,18 +601,74 @@ private boolean removeContainerInfo(String containerId) {
     return false;
   }
 
+  private Map<String, Map<String, LogEntry.Level>> copyLogLevels(Map<String, Map<String, LogEntry.Level>> originalMap) {
+    Map<String, Map<String, LogEntry.Level>> result = new TreeMap<>();
+    for (Map.Entry<String, Map<String, LogEntry.Level>> entry : originalMap.entrySet()) {
+      Map<String, LogEntry.Level> value = new TreeMap<>();
+      for (Map.Entry<String, LogEntry.Level> valueEntry : entry.getValue().entrySet()) {
+        value.put(valueEntry.getKey(), valueEntry.getValue());
+      }
+      result.put(entry.getKey(), value);
+    }
+    return result;
+  }
+
+  private void checkAndSetLogLevels(Message message, String runnableName) {
+    if (message.getType() != Message.Type.SYSTEM
+      || !SystemMessages.LOG_LEVEL.equals(message.getCommand().getCommand())) {
+      return;
+    }
+
+    Map<String, LogEntry.Level> runnableLogLevels = convertLogLevelValuesToLogEntry(message.getCommand().getOptions());
+    if (!logLevels.containsKey(runnableName)) {
+      logLevels.put(runnableName, runnableLogLevels);
+    } else {
+      logLevels.get(runnableName).putAll(runnableLogLevels);
+    }
+  }
+
+  private Location saveLogLevels() {
```

Comment:
We save the log levels to file because we have no way to know if there is message about log level change come before the launch of some containers(this can happen when we change instance count or restart the container). So when a message about log level change come, we will have it stored in the Map in `RunningContainers`. And if there is a container launch afterwards, we will check if there is update in the log levels, and localize it for the container if necessary.